File: web/rgbd-js/depth_encoder.js
import { NativeByteArray } from "./capi_containers.js";

export class NativeDepthEncoder {
  constructor(wasmModule, ptr) {
    this.wasmModule = wasmModule;
    this.ptr = ptr;
  }

  static createRVLEncoder(wasmModule, width, height) {
    const ptr = wasmModule.ccall("rgbd_depth_encoder_create_rvl_encoder",
                                 "number",
                                 ["number", "number"],
                                 [width, height]);
    return new NativeDepthEncoder(wasmModule, ptr);
  }

  static createTDC1Encoder(wasmModule, width, height, depthDiffMultiplier) {
    const ptr = wasmModule.ccall("rgbd_depth_encoder_create_tdc1_encoder",
                                 "number",
                                 ["number", "number", "number"],
                                 [width, height, depthDiffMultiplier]);
    return new NativeDepthEncoder(wasmModule, ptr);
  }

  close() {
    this.wasmModule.ccall("rgbd_depth_encoder_dtor", null, ["number"], [this.ptr]);
  }

  encode(depthValues, keyframe) {
    const depthValuesPtr = this.wasmModule._malloc(depthValues.byteLength);
    this.wasmModule.HEAPU8.set(depthValues, depthValuesPtr);
    const byteArrayPtr = this.wasmModule.ccall("rgbd_depth_encoder_encode",
                                               "number",
                                               ["number", "number", "boolean"],
                                               [this.ptr, depthValuesPtr, keyframe]);
    this.wasmModule._free(depthValuesPtr);

    const byteArray = new NativeByteArray(this.wasmModule, byteArrayPtr);
    const bytes = byteArray.toArray();
    byteArray.close();
    return bytes;
  }
}

File: python/pyrgbd/depth_encoder.py
from ._librgbd_ffi import lib
from .capi_containers import NativeByteArray
import numpy as np
from .utils import cast_np_array_to_pointer


class NativeDepthEncoder:
    def __init__(self, ptr):
        self.ptr = ptr

    @staticmethod
    def create_rvl_encoder(width: int, height: int):
        return NativeDepthEncoder(lib.rgbd_depth_encoder_create_rvl_encoder(width, height))

    @staticmethod
    def create_tdc1_encoder(width: int, height: int, depth_diff_multiplier: int):
        return NativeDepthEncoder(lib.rgbd_depth_encoder_create_tdc1_encoder(width, height, depth_diff_multiplier))

    def close(self):
        lib.rgbd_depth_encoder_dtor(self.ptr)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def encode(self, depth_values: np.ndarray, keyframe: bool) -> np.array:
        return NativeByteArray(
            lib.rgbd_depth_encoder_encode(self.ptr,
                                          cast_np_array_to_pointer(depth_values),
                                          keyframe)).to_np_array()
